rust/src/io/reader.rs

```diff
@@ -325,17 +363,35 @@ async fn read_batch(
             .ok_or_else(|| Error::IO {
                 message: format!("batch {batch_id} does not exist"),
             })?;
-        let row_id_arr = Arc::new(UInt64Array::from_iter_values(
-            ids_in_batch
-                .iter()
-                .map(|o| compute_row_id(reader.fragment_id, *o + batch_offset)),
-        ));
+        let row_ids: Vec<u64> = ids_in_batch
+            .iter()
+            .map(|o| compute_row_id(reader.fragment_id, *o + batch_offset))
+            .collect();
+
+        Some(row_ids)
+    } else {
+        None
+    };
+
+    if with_row_id {
+        let row_id_arr = Arc::new(UInt64Array::from(row_ids.clone().unwrap()));
         batch = batch.try_with_column(
             ArrowField::new("_rowid", DataType::UInt64, false),
             row_id_arr,
         )?;
     }
-    Ok(batch)
+
+    // TODO: This is a minor cop out. Pushing deletion vector in to the decoders is hard
+    // so I'm going to just leave deletion filter at this layer for now.
```

Comment:
I agree this is a decent course of action for now. I think we should plan to push this down further at the same time we do statistics-based predicate pushdown.

Reply:
added more context to TODO